File: Homework_03/KP/Python/main.py
import random
from typing import List, Tuple
from collections import namedtuple
import time

# Define the Item class
Item = namedtuple('Item', ['index', 'weight', 'value'])

class Genome:
    def __init__(self, items: List[Item], max_weight: int, chromosome: List[int] = None):
        self.items = items
        self.max_weight = max_weight
        self.chromosome = chromosome if chromosome else self.generate_feasible_chromosome()
        self.fitness = self.calculate_fitness()

    def generate_feasible_chromosome(self) -> List[int]:
        # Generate a feasible chromosome by adding items randomly until max_weight is reached
        chromosome = [0] * len(self.items)
        indices = list(range(len(self.items)))
        random.shuffle(indices)
        total_weight = 0
        for i in indices:
            if total_weight + self.items[i].weight <= self.max_weight:
                chromosome[i] = 1
                total_weight += self.items[i].weight
        return chromosome

    def calculate_fitness(self) -> float:
        total_weight = sum(item.weight for item, gene in zip(self.items, self.chromosome) if gene)
        total_value = sum(item.value for item, gene in zip(self.items, self.chromosome) if gene)
        # Penalize overweight solutions by reducing fitness proportionally to excess weight
        if total_weight > self.max_weight:
            return 0
        else:
            return total_value

class GeneticAlgorithm:
    def __init__(self, items: List[Item], max_weight: int, population_size: int, generations: int, mutation_probability: float):
        self.items = items
        self.max_weight = max_weight
        self.population_size = population_size
        self.generations = generations
        self.mutation_probability = mutation_probability
        self.population = self.initialize_population()
        self.best_fitness_history = []

    def initialize_population(self) -> List[Genome]:
        # Initialize the population with feasible chromosomes
        return [Genome(self.items, self.max_weight) for _ in range(self.population_size)]

    def tournament_selection(self, k: int = 3) -> Genome:
        # Select a genome using tournament selection
        selected = random.sample(self.population, k)
        return max(selected, key=lambda genome: genome.fitness)

    def two_point_crossover(self, parent1: Genome, parent2: Genome) -> Tuple[Genome, Genome]:
        # Perform two-point crossover
        length = len(parent1.chromosome)
        if length < 2:
            return parent1, parent2

        point1 = random.randint(1, length - 2)
        point2 = random.randint(point1 + 1, length - 1)

        child1_chromosome = (
            parent1.chromosome[:point1] + parent2.chromosome[point1:point2] + parent1.chromosome[point2:]
        )
        child2_chromosome = (
            parent2.chromosome[:point1] + parent1.chromosome[point1:point2] + parent2.chromosome[point2:]
        )

        child1 = Genome(self.items, self.max_weight, chromosome=child1_chromosome)
        child2 = Genome(self.items, self.max_weight, chromosome=child2_chromosome)
        return child1, child2

    def mutate(self, genome: Genome) -> None:
        # Mutate the genome by flipping bits with a set probability, enforcing feasibility
        for i in range(len(genome.chromosome)):
            if random.random() < self.mutation_probability:
                # Attempt to flip the bit
                if genome.chromosome[i] == 0:
                    # Turn on the item if feasible
                    if self.calculate_weight(genome.chromosome) + self.items[i].weight <= self.max_weight:
                        genome.chromosome[i] = 1
                else:
                    # Turn off the item
                    genome.chromosome[i] = 0
        genome.fitness = genome.calculate_fitness()

    def calculate_weight(self, chromosome: List[int]) -> int:
        # Calculate the total weight of a chromosome
        return sum(item.weight for item, gene in zip(self.items, chromosome) if gene)

    def evolve_population_with_elitism(self, elite_size: int = 2) -> None:
        # Create a new population by retaining elites and using selection, crossover, and mutation
        new_population = sorted(self.population, key=lambda genome: genome.fitness, reverse=True)[:elite_size]
        while len(new_population) < self.population_size:
            parent1 = self.tournament_selection()
            parent2 = self.tournament_selection()
            child1, child2 = self.two_point_crossover(parent1, parent2)
            self.mutate(child1)
            self.mutate(child2)
            new_population.extend([child1, child2])
        self.population = sorted(new_population, key=lambda genome: genome.fitness, reverse=True)[:self.population_size]
        self.best_fitness_history.append(self.population[0].fitness)

    def run_evolution(self, fitness_threshold: float = 1e-6, patience: int = 25) -> Genome:
        # Run the genetic algorithm with stopping condition based on fitness improvement
        fitness_improvements = []
        avg_fitness_change = None

        for generation in range(self.generations):
            self.evolve_population_with_elitism()

            if generation > 0:
                fitness_change = abs(self.best_fitness_history[-1] - self.best_fitness_history[-2])
                fitness_improvements.append(fitness_change)

                if len(fitness_improvements) > patience:
                    fitness_improvements.pop(0)

                avg_fitness_change = sum(fitness_improvements) / len(fitness_improvements)

                if avg_fitness_change < fitness_threshold and len(fitness_improvements) == patience:
                    print(f"Stopping early at generation {generation} due to minimal improvement.")
                    break

            if generation % 10 == 0:
                avg_fitness_change_str = f"{avg_fitness_change:.6f}" if avg_fitness_change else "N/A"
                print(f"Generation {generation}: Best fitness = {self.best_fitness_history[-1]}, Avg fitness change = {avg_fitness_change_str}")

        # Ensure the best genome found is feasible
        feasible_population = [genome for genome in self.population if self.calculate_weight(genome.chromosome) <= self.max_weight]
        return max(feasible_population, key=lambda genome: genome.fitness) if feasible_population else self.population[0]

def load_items(filename: str) -> Tuple[int, List[Item]]:
    # Load items from a file
    with open(filename, 'r') as file:
        lines = file.readlines()
    max_weight, num_items = map(int, lines[0].split())
    items = [Item(index=idx, weight=int(line.split()[0]), value=int(line.split()[1])) for idx, line in enumerate(lines[1:])]
    return max_weight, items

def main():
    option = input("Enter short/long: ")
    start_time = time.time()
    max_weight, items = load_items(f"../TestData/{option}_test_data")

    ga = GeneticAlgorithm(
        items=items,
        max_weight=max_weight,
        population_size=1000,
        generations=1000,
        mutation_probability=0.05
    )

    best_genome = ga.run_evolution()

    # Output the best solution found
    selected_items = [item for item, gene in zip(items, best_genome.chromosome) if gene]
    total_weight = sum(item.weight for item in selected_items)
    total_value = sum(item.value for item in selected_items)

    print(f"Total value: {total_value}")
    print(f"Total weight: {total_weight}")
    print(f"Selected items: {[item.index for item in selected_items]}")
    print(f"Time taken: {time.time() - start_time:.2f} seconds")

if __name__ == "__main__":
    main()

File: Homework_03/KP/Go/main.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
)

// Item represents an item with weight and value.
type Item struct {
	Index  int
	Weight int
	Value  int
}

// Genome represents a candidate solution in the Knapsack problem.
type Genome struct {
	Chromosome []int
	Fitness    float64
}

// GeneticAlgorithm manages the GA process.
type GeneticAlgorithm struct {
	Items               []Item
	MaxWeight           int
	PopulationSize      int
	Generations         int
	Population          []Genome
	MutationProbability float64
	BestFitnessHistory  []float64
}

// Initialize the population with feasible solutions.
func (ga *GeneticAlgorithm) InitializePopulation() {
	ga.Population = make([]Genome, ga.PopulationSize)
	for i := 0; i < ga.PopulationSize; i++ {
		chromosome := ga.generateFeasibleChromosome()
		genome := Genome{Chromosome: chromosome}
		genome.CalculateFitness(ga.Items, ga.MaxWeight)
		ga.Population[i] = genome
	}
}

// Calculate the fitness of a genome by summing values and checking weight.
func (g *Genome) CalculateFitness(items []Item, maxWeight int) {
	totalWeight := 0
	totalValue := 0.0
	for i, gene := range g.Chromosome {
		if gene == 1 {
			totalWeight += items[i].Weight
			totalValue += float64(items[i].Value)
		}
	}
	if totalWeight <= maxWeight {
		g.Fitness = totalValue
	} else {
		g.Fitness = 0 // Infeasible solutions have zero fitness.
	}
}

// Generate a feasible chromosome by adding items until maxWeight is reached.
func (ga *GeneticAlgorithm) generateFeasibleChromosome() []int {
	chromosome := make([]int, len(ga.Items))
	indices := rand.Perm(len(ga.Items))
	totalWeight := 0
	for _, idx := range indices {
		if totalWeight+ga.Items[idx].Weight <= ga.MaxWeight {
			chromosome[idx] = 1
			totalWeight += ga.Items[idx].Weight
		} else {
			chromosome[idx] = 0
		}
	}
	return chromosome
}

// Tournament selection for stronger selection pressure.
func (ga *GeneticAlgorithm) TournamentSelection(tournamentSize int) Genome {
	best := ga.Population[rand.Intn(ga.PopulationSize)]
	for i := 1; i < tournamentSize; i++ {
		contender := ga.Population[rand.Intn(ga.PopulationSize)]
		if contender.Fitness > best.Fitness {
			best = contender
		}
	}
	return best
}

// Two-point crossover to generate diverse offspring.
func (ga *GeneticAlgorithm) TwoPointCrossover(parent1, parent2 Genome) (Genome, Genome) {
	child1 := Genome{Chromosome: make([]int, len(parent1.Chromosome))}
	child2 := Genome{Chromosome: make([]int, len(parent2.Chromosome))}

	point1 := rand.Intn(len(parent1.Chromosome))
	point2 := rand.Intn(len(parent1.Chromosome))
	if point1 > point2 {
		point1, point2 = point2, point1
	}

	// Copy genes between point1 and point2 from each parent
	copy(child1.Chromosome[point1:point2], parent1.Chromosome[point1:point2])
	copy(child2.Chromosome[point1:point2], parent2.Chromosome[point1:point2])

	// Fill in remaining genes from the other parent, maintaining order
	for i := 0; i < len(parent1.Chromosome); i++ {
		if i < point1 || i >= point2 {
			child1.Chromosome[i] = parent2.Chromosome[i]
			child2.Chromosome[i] = parent1.Chromosome[i]
		}
	}

	child1.CalculateFitness(ga.Items, ga.MaxWeight)
	child2.CalculateFitness(ga.Items, ga.MaxWeight)
	return child1, child2
}

// Mutation function that ensures feasible mutations.
func (ga *GeneticAlgorithm) Mutate(genome *Genome) {
	for i := range genome.Chromosome {
		if rand.Float64() < ga.MutationProbability {
			curr_weight := ga.calculateWeight(genome.Chromosome)
			if genome.Chromosome[i] == 0 {
				// Turn on the item if feasible
				if curr_weight+ga.Items[i].Weight <= ga.MaxWeight {
					genome.Chromosome[i] = 1
				}
			} else {
				// Turn off the item
				genome.Chromosome[i] = 0
				curr_weight -= ga.Items[i].Weight
			}
		}
	}
	genome.CalculateFitness(ga.Items, ga.MaxWeight)
}

// Calculate the total weight of the chromosome.
func (ga *GeneticAlgorithm) calculateWeight(chromosome []int) int {
	totalWeight := 0
	for i, gene := range chromosome {
		if gene == 1 {
			totalWeight += ga.Items[i].Weight
		}
	}
	return totalWeight
}

// Evolve Population with Elitism and Tournament Selection.
func (ga *GeneticAlgorithm) EvolvePopulationWithElitism(eliteSize, tournamentSize int) {
	newPopulation := make([]Genome, 0, ga.PopulationSize)
	sort.Slice(ga.Population, func(i, j int) bool {
		return ga.Population[i].Fitness > ga.Population[j].Fitness
	})
	// Retain elite genomes
	newPopulation = append(newPopulation, ga.Population[:eliteSize]...)

	for len(newPopulation) < ga.PopulationSize {
		parent1 := ga.TournamentSelection(tournamentSize)
		parent2 := ga.TournamentSelection(tournamentSize)
		child1, child2 := ga.TwoPointCrossover(parent1, parent2)
		ga.Mutate(&child1)
		ga.Mutate(&child2)
		newPopulation = append(newPopulation, child1, child2)
	}

	if len(newPopulation) > ga.PopulationSize {
		newPopulation = newPopulation[:ga.PopulationSize]
	}
	ga.Population = newPopulation
	ga.BestFitnessHistory = append(ga.BestFitnessHistory, ga.Population[0].Fitness)
}

// Run the genetic algorithm evolution process.
func (ga *GeneticAlgorithm) RunEvolution(fitnessThreshold float64, patience int) Genome {
	fitnessImprovements := []float64{}
	for generation := 0; generation < ga.Generations; generation++ {
		ga.EvolvePopulationWithElitism(5, 3)

		if generation > 0 {
			fitnessChange := mathAbs(ga.BestFitnessHistory[len(ga.BestFitnessHistory)-1] - ga.BestFitnessHistory[len(ga.BestFitnessHistory)-2])
			fitnessImprovements = append(fitnessImprovements, fitnessChange)

			if len(fitnessImprovements) > patience {
				fitnessImprovements = fitnessImprovements[1:]
			}

			avgFitnessChange := averageFloats(fitnessImprovements)
			if avgFitnessChange < fitnessThreshold && len(fitnessImprovements) == patience {
				fmt.Printf("Stopping early at generation %d due to minimal improvement.\n", generation)
				break
			}
		}

		if generation%10 == 0 {
			fmt.Printf("Generation %d: Best fitness = %.2f\n", generation, ga.BestFitnessHistory[len(ga.BestFitnessHistory)-1])
		}
	}
	return ga.Population[0]
}

// Utility Functions
func mathAbs(a float64) float64 {
	if a < 0 {
		return -a
	}
	return a
}

func averageFloats(a []float64) float64 {
	sum := 0.0
	for _, v := range a {
		sum += v
	}
	return sum / float64(len(a))
}

func loadItems(filename string) (int, []Item, error) {
	file, err := os.Open(filename)
	if err != nil {
		return 0, nil, err
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	var items []Item
	maxWeight := 0
	numItems := 0
	idx := 0
	if scanner.Scan() {
		line := scanner.Text()
		parts := strings.Fields(line)
		if len(parts) != 2 {
			return 0, nil, fmt.Errorf("Invalid first line in file")
		}
		maxWeight, _ = strconv.Atoi(parts[0])
		numItems, _ = strconv.Atoi(parts[1])
	}
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Fields(line)
		weight, _ := strconv.Atoi(parts[0])
		value, _ := strconv.Atoi(parts[1])
		item := Item{
			Index:  idx,
			Weight: weight,
			Value:  value,
		}
		items = append(items, item)
		idx++
	}
	if len(items) != numItems {
		return 0, nil, fmt.Errorf("Number of items does not match")
	}
	return maxWeight, items, nil
}

func main() {
	rand.Seed(time.Now().UnixNano())
	fmt.Print("Enter 'short' or 'long': ")
	var option string
	fmt.Scan(&option)
	startTime := time.Now()
	maxWeight, items, err := loadItems(fmt.Sprintf("../TestData/%s_test_data", option))
	if err != nil {
		fmt.Println("Error loading items:", err)
		return
	}

	ga := GeneticAlgorithm{
		Items:               items,
		MaxWeight:           maxWeight,
		PopulationSize:      10000,
		Generations:         2000,
		MutationProbability: 0.15,
	}
	ga.InitializePopulation()
	bestGenome := ga.RunEvolution(1e-6, 25)

	// Output the best solution found
	selectedItems := []Item{}
	totalWeight := 0
	totalValue := 0.0
	for i, gene := range bestGenome.Chromosome {
		if gene == 1 {
			selectedItems = append(selectedItems, items[i])
			totalWeight += items[i].Weight
			totalValue += float64(items[i].Value)
		}
	}

	fmt.Printf("Total value: %.0f\n", totalValue)
	fmt.Printf("Total weight: %d\n", totalWeight)
	fmt.Printf("Selected items indices: ")
	for _, item := range selectedItems {
		fmt.Printf("%d ", item.Index)
	}
	fmt.Printf("\nTime taken: %.2f seconds\n", time.Since(startTime).Seconds())
}
